oauth2/src/main/java/com/linecorp/armeria/common/auth/oauth2/AbstractOAuth2Request.java

```diff
@@ -0,0 +1,224 @@
+    private final String endpointPath;
+    @Nullable
+    private final ClientAuthorization clientAuthorization;
+
+    /**
+     * A common abstraction for the requests implementing various Authorization request/response flows,
+     * as per <a href="https://tools.ietf.org/html/rfc6749">[RFC6749]</a> and other relevant specifications.
+     *
+     * @param endpoint A {@link WebClient} to facilitate the Authorization requests. Must
+     *                 correspond to the required Authorization endpoint of the OAuth 2 system.
+     * @param endpointPath A URI path that corresponds to the Authorization endpoint of the
+     *                     OAuth 2 system.
+     * @param clientAuthorization Provides client authorization for the OAuth requests,
+     *                            as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     */
+    protected AbstractOAuth2Request(WebClient endpoint, String endpointPath,
+                                    @Nullable ClientAuthorization clientAuthorization) {
+        this.endpoint = Objects.requireNonNull(endpoint, "endpoint");
+        this.endpointPath = Objects.requireNonNull(endpointPath, "endpointPath");
+        this.clientAuthorization = clientAuthorization; // optional
+    }
+
+    /**
+     * Returns the {@link WebClient} of the authorization endpoint.
+     */
+    protected WebClient endpoint() {
+        return endpoint;
+    }
+
+    /**
+     * Returns the authorization endpoint path.
+     */
+    protected String endpointPath() {
+        return endpointPath;
+    }
+
+    /**
+     * Returns the client authorization object.
+     */
+    @Nullable
+    protected ClientAuthorization clientAuthorization() {
+        return clientAuthorization;
+    }
+
+    /**
+     * Extracts data from OK response and converts it to the target type {@code T}.
+     */
+    protected abstract T extractOkResults(AggregatedHttpResponse response, Map<String, String> requestData);
+
+    /**
+     * Returns the value for the {@link HttpHeaderNames#AUTHORIZATION}.
+     */
+    @Nullable
+    protected String authorizationHeaderValue() {
+        return clientAuthorization == null ? null : clientAuthorization.authorizationHeaderValue();
+    }
+
+    /**
+     * Sets client credentials as form data parameters,
+     * as per <a href="https://tools.ietf.org/html/rfc6749#section-2.3">[RFC6749], Section 2.3</a>.
+     */
+    protected void setCredentialsAsBodyParameters(QueryParamsBuilder formBuilder) {
+        if (clientAuthorization != null) {
+            clientAuthorization.setCredentialsAsBodyParameters(formBuilder);
+        }
+    }
+
+    /**
+     * Makes a request to the authorization endpoint using supplied {@code requestForm} parameters and converts
+     * the result to the given type {@code T}.
+     */
+    protected CompletableFuture<T> executeWithParameters(LinkedHashMap<String, String> requestFormData) {
```

Comment:
Is it possible to use `QueryParams` instead of `Map<String, String>` here?

Reply:
Fixed